src/cli/cli.cpp

```diff
@@ -8251,256 +8251,357 @@ otError Interpreter::ProcessCommand(Arg aArgs[])
     }
 
     static constexpr Command kCommands[] = {
-#if OPENTHREAD_FTD || OPENTHREAD_MTD
-#if OPENTHREAD_CONFIG_BORDER_AGENT_ENABLE
+#ifdef CMD_ba
         CmdEntry("ba"),
 #endif
-#if (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
+#ifdef CMD_bbr
         CmdEntry("bbr"),
 #endif
-#if OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE
+#ifdef CMD_br
         CmdEntry("br"),
 #endif
+#ifdef CMD_bufferinfo
         CmdEntry("bufferinfo"),
+#endif
+#ifdef CMD_ccathreshold
         CmdEntry("ccathreshold"),
-#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+#endif
+#ifdef CMD_ccm
         CmdEntry("ccm"),
 #endif
+#ifdef CMD_channel
         CmdEntry("channel"),
-#if OPENTHREAD_FTD
+#endif
+#ifdef CMD_child
         CmdEntry("child"),
+#endif
+#ifdef CMD_childip
         CmdEntry("childip"),
+#endif
+#ifdef CMD_childmax
         CmdEntry("childmax"),
+#endif
+#ifdef CMD_childrouterlinks
         CmdEntry("childrouterlinks"),
 #endif
+#ifdef CMD_childsupervision
         CmdEntry("childsupervision"),
+#endif
+#ifdef CMD_childtimeout
         CmdEntry("childtimeout"),
-#if OPENTHREAD_CONFIG_COAP_API_ENABLE
+#endif
+#ifdef CMD_coap
         CmdEntry("coap"),
 #endif
-#if OPENTHREAD_CONFIG_COAP_SECURE_API_ENABLE
+#ifdef CMD_coaps
         CmdEntry("coaps"),
 #endif
-#if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE
+#ifdef CMD_coex
         CmdEntry("coex"),
 #endif
-#if OPENTHREAD_CONFIG_COMMISSIONER_ENABLE && OPENTHREAD_FTD
+#ifdef CMD_commissioner
         CmdEntry("commissioner"),
 #endif
-#if OPENTHREAD_FTD
+#ifdef CMD_contextreusedelay
         CmdEntry("contextreusedelay"),
 #endif
+#ifdef CMD_counters
         CmdEntry("counters"),
-#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+#endif
+#ifdef CMD_csl
         CmdEntry("csl"),
 #endif
+#ifdef CMD_dataset
         CmdEntry("dataset"),
+#endif
+#ifdef CMD_debug
         CmdEntry("debug"),
-#if OPENTHREAD_FTD
+#endif
+#ifdef CMD_delaytimermin
         CmdEntry("delaytimermin"),
 #endif
+#ifdef CMD_detach
         CmdEntry("detach"),
-#endif // OPENTHREAD_FTD || OPENTHREAD_MTD
-#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_MLE_DEVICE_PROPERTY_LEADER_WEIGHT_ENABLE
+#endif
+#ifdef CMD_deviceprops
         CmdEntry("deviceprops"),
 #endif
-#if OPENTHREAD_CONFIG_DIAG_ENABLE
+#ifdef CMD_diag
         CmdEntry("diag"),
 #endif
-#if OPENTHREAD_FTD || OPENTHREAD_MTD
+#ifdef CMD_discover
         CmdEntry("discover"),
-#if OPENTHREAD_CONFIG_DNS_CLIENT_ENABLE || OPENTHREAD_CONFIG_DNSSD_SERVER_ENABLE || \
-    OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+#endif
+#ifdef CMD_dns
```

Comment:
Should add similar check to the definitions of the `Process<Cmd("xyz">)`?

For example, instead of 
```c++
#if OPENTHREAD_CLI_DNS_ENABLE
template <> otError Interpreter::Process<Cmd("dns")>(Arg aArgs[]) { return mDns.Process(aArgs); }
#endif
```
we can use
```c++
#ifdef CMD_dns
template <> otError Interpreter::Process<Cmd("dns")>(Arg aArgs[]) { return mDns.Process(aArgs); }
#endif
```

Not sure if this would be better or if we should keep the original model. If we do this, it would be consistent in terms of using `#if` checks to decide about commands, but then it may add a lot of extra `#ifdef` to the code.